src/main/java/in/projecteka/consentmanager/user/TransactionPinService.java

```diff
@@ -58,25 +60,38 @@ private boolean isPinValid(String pin) {
                 .switchIfEmpty(Mono.error(ClientError.requestAlreadyExists()))
                 .flatMap(val -> transactionPinRepository.updateRequestId(requestId, patientId)
                         .then(transactionPinRepository.getTransactionPinByPatient(patientId)
-                        .flatMap(transactionPin -> {
-                            if (transactionPin.isEmpty()) {
-                                return Mono.error(ClientError.transactionPinNotFound());
-                            }
-                            if (!encoder.matches(pin, transactionPin.get().getPin())) {
-                                return Mono.error(ClientError.transactionPinDidNotMatch());
-                            }
-                            return Mono.empty();
-                        }).thenReturn(newToken(patientId, scope))));
+                                .flatMap(transactionPin -> {
+                                    if (transactionPin.isEmpty()) {
+                                        return Mono.error(ClientError.transactionPinNotFound());
+                                    }
+                                    return dayCache.exists(blockedKey(patientId))
+                                            .filter(exists -> !exists)
+                                            .switchIfEmpty(Mono.error(ClientError.invalidAttemptsExceeded()))
+                                            .then(Mono.defer(()-> Mono.just(encoder.matches(pin, transactionPin.get().getPin()))))
+                                            .filter(matches-> !matches)
+                                            .flatMap(doesNotMatch -> dayCache.increment(incorrectAttemptKey(patientId))
+                                                    .filter(count-> count != userServiceProperties.getMaxIncorrectPinAttempts())
+                                                    .switchIfEmpty(Mono.defer(() -> dayCache.put(blockedKey(patientId),"true").thenReturn(userServiceProperties.getMaxIncorrectPinAttempts())))
+                                                    .flatMap(attemptCount -> Mono.error(ClientError.transactionPinDidNotMatch(String.format("%s attempts left",userServiceProperties.getMaxIncorrectPinAttempts() - attemptCount)))));
+                                }).thenReturn(newToken(patientId, scope))));
```

Comment:
Can we avoid this nested `filter`, `flatMap` chains?

Reply:
In the scope of the current task, this nesting can be justified. As in, I had to have nesting based on the preexisting code.

Reply:
Ok. But for an example
```
if (transactionPin.isEmpty()) {
                                        return Mono.error(ClientError.transactionPinNotFound());
                                    }
```
It can be taken out using a `filter` and then using a `switchIfEmpty` to throw error. Something like this:-

```
.filter(TransactionPin::isEmpty)
.switchIfEmpty(Mono.defer(() -> Mono.error(ClientError.transactionPinNotFound())))
.flatMap(transactionPin -> {
                                    return dayCache.exists(blockedKey(patientId))
                                            .filter(exists -> !exists)
                                            .switchIfEmpty(Mono.error(ClientError.invalidAttemptsExceeded()))
                                            .then(Mono.defer(()-> Mono.just(encoder.matches(pin, transactionPin.get().getPin()))))
                                            .filter(matches-> !matches)
                                            .flatMap(doesNotMatch -> dayCache.increment(incorrectAttemptKey(patientId))
                                                    .filter(count-> count != userServiceProperties.getMaxIncorrectPinAttempts())
                                                    .switchIfEmpty(Mono.defer(() -> dayCache.put(blockedKey(patientId),"true").thenReturn(userServiceProperties.getMaxIncorrectPinAttempts())))
                                                    .flatMap(attemptCount -> Mono.error(ClientError.transactionPinDidNotMatch(String.format("%s attempts left",userServiceProperties.getMaxIncorrectPinAttempts() - attemptCount)))));
                                })

Reply:
Sorry for bad formatting. It just made it like that.

Reply:
Updated